base/compiler/typeinfer.jl

```diff
@@ -229,34 +229,32 @@ function finish!(interp::AbstractInterpreter, caller::InferenceState;
     if opt isa OptimizationState
         result.src = opt = ir_to_codeinf!(opt)
     end
-    if isdefined(result, :ci)
+    if isdefined(result, :ci) # implies `is_cached(caller)`
```

Comment:
This is only true by accident right now. It will be false soon (https://github.com/JuliaLang/julia/pull/54816)